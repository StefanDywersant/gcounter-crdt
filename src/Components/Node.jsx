import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Beans from './Beans';
import Bean from './Bean';
import Button from './Button';
import styles from './Node.scss';


/**
 * Renders bean counter node.
 */
class Node extends Component {

	/**
	 * Constructor
	 * @param {{total: number, onIncrementClick: function}} props Component's properties
	 */
	constructor(props) {
		super(props);
		this.state = {
			beans: []
		};
	}

	/**
	 * Current bean socket absolute position.
	 * @returns {{x: number, y: number}|null}
	 */
	get socketPosition() {
		if (!this.el) {
			return null;
		}

		const box = this.el.getBoundingClientRect();

		const body = document.body;

		const clientTop = body.clientTop || 0;
		const clientLeft = body.clientLeft || 0;

		const top  = box.top +  window.pageYOffset - clientTop;
		const left = box.left + window.pageXOffset - clientLeft;

		return {
			y: Math.round(top + 10),
			x: Math.round(left + 20)
		};
	}

	/**
	 * Invoked before component receives props.
	 * @param {object} nextProps New properties
	 */
	componentWillReceiveProps(nextProps) {
		if (nextProps.total !== this.state.beans.length) {
			// recalculate beans position in order to trigger "bean push" animation
			this.setState({
				beans: new Array(nextProps.total)
					.fill(0)
					.map(() => ({slideX: Math.random(), slideY: Math.random()}))
			});
		}
	}

	/**
	 * Renders node box to DOM.
	 * @returns {XML}
	 */
	render() {
		return <div
			className={[styles.container, this.props.className].join(' ')}
			ref={el => this.el = el}
		>
			<Beans className={styles.beans}>
				{this.state.beans.map((bean, i) => <Bean key={i} {...bean} />)}
			</Beans>

			<p className={styles.total}>
				Total: <strong>{this.props.total}</strong>
			</p>

			<Button
				className={styles.button}
				onClick={this.props.onIncrementClick}
			>
				+increment
			</Button>
		</div>
	}

}


Node.propTypes = {

	/**
	 * Total beans count
	 */
	total: PropTypes.number.isRequired,

	/**
	 * Increment button handler
	 */
	onIncrementClick: PropTypes.func

};


export default Node;