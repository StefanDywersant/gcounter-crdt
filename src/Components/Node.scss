.container {
	border: 1px dotted #004887;
	border-radius: 20px;
	background: rgba(0, 140, 254, 0.1);
	padding: 10px 20px;
	width: 100px;

	display: flex;
	flex-direction: column;
	align-items: center;
}

.beans {
	width: 100%;
	min-height: 100px;
}

.total {
	text-align: center;
	text-transform: uppercase;
	width: 100%;
}

.button {
	margin: 10px;
}