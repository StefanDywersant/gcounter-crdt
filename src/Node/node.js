/**
 * Minimum time which must pass before sending state update to other node
 * @type {number}
 */
const DELAY_MIN = 1000;

/**
 * Maximum time which may pass before sending state update to other node
 * @type {number}
 */
const DELAY_MAX = 5000;

/**
 * Self node identifier
 * @type {number|null}
 */
let myId = null;

/**
 * Node's internal state
 * @type {number[]|null}
 */
let myState = null;


/**
 * Initializes node
 * @param {number} id Node identifier
 * @param {number} nodes All node count
 */
const init = function(id, nodes) {
	myId = id;
	myState = new Array(nodes).fill(0);
};

/**
 * Increments local counter and broadcasts the update to other nodes.
 */
const increment = function() {
	if (!myState) {
		throw new Error('Worker not initialized');
	}

	myState[myId]++;

	// broadcast state copy to other nodes
	broadcast(myState.concat());
};

/**
 * Merges foreign state into local state.
 * @param {number[]} state Foreign state
 */
const merge = function(state) {
	if (!myState) {
		throw new Error('Worker not initialized');
	}

	state.forEach((value, i) => myState[i] = Math.max(value, myState[i]));

	// send UI update immediately
	self.postMessage({
		type: 'ui:total',
		total: myState.reduce((total, node) => total + node, 0)
	});
};

/**
 * Broadcasts state update to all other nodes.
 * @param {number[]} state State being broadcasted
 */
const broadcast = function(state) {
	state.forEach((_, id) => {
		// do not send update to self
		if (id === myId) {
			return;
		}

		const delay = Math.floor(DELAY_MIN + Math.random() * (DELAY_MAX - DELAY_MIN));

		setTimeout(() => {
			self.postMessage({
				type: 'state',
				target: id,
				state
			});
		}, delay);

		// send UI update immediately
		self.postMessage({
			type: 'ui:communication',
			target: id,
			duration: delay
		});
	});
};

// handles incoming message
self.onmessage = (event) => {
	const message = event.data;

	switch (message.type) {
		case 'init':
			init(message.id, message.nodes);
			break;

		case 'increment':
			increment();
			break;

		case 'state':
			merge(message.state);
			break;
	}
};