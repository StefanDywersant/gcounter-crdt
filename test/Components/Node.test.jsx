import React from 'react';
import ReactShallowRenderer from 'react-test-renderer/shallow';
import Node from "../../src/Components/Node";

const renderer = new ReactShallowRenderer();

describe('Component rendering', () => {

	it('renders correctly with default props', () => {
		expect(renderer.render(
			<Node total={32} />
		)).toMatchSnapshot();
	});

});
