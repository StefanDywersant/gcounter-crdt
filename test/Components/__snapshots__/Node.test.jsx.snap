// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component rendering renders correctly with default props 1`] = `
<div
  className="container "
>
  <Beans
    className="beans"
  >
    
  </Beans>
  <p
    className="total"
  >
    Total: 
    <strong>
      32
    </strong>
  </p>
  <Button
    className="button"
    onClick={undefined}
  >
    +increment
  </Button>
</div>
`;
